pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -1531,3 +1532,80 @@ def _handle_distribution_tree(declarative_content):
+    units inside Pulp with the same unit key. Any existing Content objects found, replace their
+    "unsaved" counterpart in the :class:`~pulpcore.plugin.stages.DeclarativeContent` object.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `self._out_q` after it has
+    been handled.
+
+    This stage drains all available items from `self._in_q` and batches everything into one large
+    call to the db for efficiency.
+    """
+
+    async def run(self):
+        """
+        The coroutine for this stage.
+
+        Returns:
+            The coroutine for this stage.
+        """
+        async for batch in self.batches():
+            content_q_by_type = defaultdict(lambda: Q(pk__in=[]))
+            d_content_by_nat_key = defaultdict(list)
+            for d_content in batch:
+                if d_content.content._state.adding:
+                    model_type = type(d_content.content)
+                    unit_q = d_content.content.q()
+                    content_q_by_type[model_type] = content_q_by_type[model_type] | unit_q
+                    d_content_by_nat_key[d_content.content.natural_key()].append(d_content)
+
+            for model_type, content_q in content_q_by_type.items():
+                try:
+                    await sync_to_async(model_type.objects.filter(content_q).touch)()
+                except AttributeError:
+                    raise TypeError(
+                        "Plugins which declare custom ORM managers on their content classes "
+                        "should have those managers inherit from "
+                        "pulpcore.plugin.models.ContentManager."
+                    )
+                async for result in sync_to_async_iterable(
+                    model_type.objects.filter(content_q).iterator()
+                ):
+                    for d_content in d_content_by_nat_key[result.natural_key()]:
+                        # ============ The below lines are added ============
+                        if model_type == Package:
+                            incorrect_changelogs = result.changelogs != d_content.content.changelogs
+                            incorrect_modular_relation = (
+                                not result.is_modular and d_content.content.is_modular
+                            )
+
+                            if incorrect_changelogs:
+                                # Covers a class of issues with changelog data on the CDN
+                                result.changelogs = d_content.content.changelogs
+                            if incorrect_modular_relation:
+                                # Covers #2643
+                                result.is_modular = True
+
+                            if incorrect_changelogs or incorrect_modular_relation:
+                                log.info("Updated data for package {}".format(result.nevra()))
+                                result.save()
```

Comment:
QueryExistingContents does not (in spite of its doc) call save(). Being up to my eyebrows right now in questions about "what happens when multiple concurrent syncs are syncing overlapping content at the same time", and the resulting Pain, I'd like to avoid any more potential deadlocks/db-exceptions. Is there a reason to require save() here?

Reply:
@ggainey If you look at the next stage, it assumes that any content units that are already saved don't need to be re-saved, and this stage is replacing the non-saved units with saved units, so even if we make changes to the object it isn't going to be re-saved later on.

We could possibly do a `select_for_update()`?

Reply:
> https://github.com/pulp/pulpcore/blob/main/pulpcore/plugin/stages/content_stages.py#L112-L113

Yeah that is exactly the start of the code-block that is causing the aforementioned Pain.

Looking at this code again, my understanding is something like this:

We find the content-units we are Adding (1571-1575), look for any that are Package (1591), fix any problems (1592-1602), and save() the result (1606). Regardless of whether we updated-in-db the unit, we replace the existing one in the DeclarativeContent entity with the one we read out of the DB (1608), which is therefore **not** marked as "adding", and therefore will be marked by the ContentSaver code at https://github.com/pulp/pulpcore/blob/main/pulpcore/plugin/stages/content_stages.py#L114 as already-saved and cause us to skip the "if" at 114 in that code. 

That "if" exists to save content and handle errors when there's a collision with Some Other Thread - see https://github.com/pulp/pulpcore/blob/main/pulpcore/plugin/stages/content_stages.py#L118. However, in our case, there's no chance of an IntegrityViolation in this thread. The IntegrityViolation happens if two threads decide to save a unit with the same unique-keys **but different pulp-ids** , because one or both is trying to create the unit. In this case, we're only here because the unit exists in the db already, and we know it, so we're using the existing-unit's pulp-id, and we save into The Right Thing. Whew.

I...*think* this all hangs together? But it's def an arcane bit of codepath. 

RE select_for_update() - that's only useful when we're about to do bulk-operations, isn't without its own dangers, and doesn't universally protect us from deadlocks. As I am currently beating my head against.

Reply:
@ggainey  This is how I read this too.
I think it is ok to save() on already saved in the db unit. The `_state.adding` check is used only in these 2 mentioned places.